### PR TITLE
Refactoring mouse events.

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description

This pull request (PR) introduces a new approach to handle mouse events. Previously, mouse events were managed in the following manner:

- All mouse events were directed to the `Scene.mouse.frp` structure, which was a significant FRP multiplexer for the events.
- Whenever a new shape was under the mouse, events were redirected from the aforementioned structure to the `PointerTarget` structure, which was included in every shape defined with the `ensogl_core::shape!` macro.
- Users of EnsoGL could read the events either from the `shape.events` structure or the `Scene.mouse.frp` structure.

This API is preserved as refactoring all of its usages is a complex task that should be performed gradually. The API was renamed to discourage its further use. In particular, the following renaming was applied:

- `Scene.mouse.frp` -> `Scene.mouse.frp_deprecated`;
- `shape.events` -> `shape.events_deprecated`;
- `PointerTarget` -> `PointerTarget_DEPRECATED`.

One significant issue with the above API was that the events were not propagated through the display object hierarchy, making it impossible to handle them in a flexible way. For instance, to implement a vector editor, some events need to be caught on a shape's parent and then decide whether to propagate them to the children, such as when dragging sliders belonging to the vector editor.

Therefore, the above APIs were replaced by the event system built into the display objects API. For instance, instead of using `shape.events.mouse_down`, you can now use `shape.on_event::<mouse::Down>()`. This provides you with a similar FRP but a more powerful endpoint as it allows:

- Controlling the event propagation (capturing and bubbling phases);
- Cancelling the event;
- Reading additional data from the event, such as the associated position on the screen of the pointer when the event was emitted (similar to JS events).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: lib/rust/ensogl/core/src/control/io/mouse/event.rs [hunk 1]
```diff
@@ -14,90 +14,244 @@ use web::dom::Shape;
...
+{
+    /// Constructor.
+    pub fn new(js_event: JsEvent, shape: Shape) -> Self {
+        let js_event = Some(js_event);
+        let event_type = default();
+        Self { js_event, shape, event_type }
+    }
+
+    /// The X coordinate of the mouse pointer relative to the position of the padding edge of the
+    /// target node.
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/control/io/mouse/event.rs]: What is `target node`? Is it `display::Object` instance? A link to proper structure would be helpful.

File: lib/rust/ensogl/core/src/control/io/mouse/event.rs [hunk 2]
```diff
@@ -14,90 +14,244 @@ use web::dom::Shape;
...
-        impl AsRef<web::Event> for $name {
-            fn as_ref(&self) -> &web::Event {
-                let js_event = AsRef::<web::$js_event>::as_ref(self);
-                js_event.as_ref()
-            }
+    /// The Y coordinate of the mouse pointer in local (DOM content) coordinates.
+    pub fn client_y(&self) -> i32 {
+        self.shape.height as i32
+            - self.js_event.as_ref().map(|t| t.as_ref().client_y()).unwrap_or_default()
+    }
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/control/io/mouse/event.rs]: Something is strange here. `client_y` and `offset_y` are relative to different things (as documentation implies), but we subtract the original JS value from the same `self.shape.height`.

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 3]
```diff
@@ -120,25 +126,32 @@ impl PointerTargetRegistry {
...
-    pub scene_frp:     Frp,
+    pub mouse_manager:   MouseManager,
+    pub last_position:   Rc<Cell<Vector2<i32>>>,
+    pub position:        Uniform<Vector2<i32>>,
+    pub click_count:     Uniform<i32>,
+    pub hover_rgba:      Uniform<Vector4<u32>>,
+    pub target:          Rc<Cell<PointerTargetId>>,
+    pub handles:         Rc<[callback::Handle; 6]>,
+    pub scene_frp:       Frp,
+    pub last_move_event: Rc<RefCell<Option<mouse::Move>>>,
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/scene.rs]: Why do we remember last move event? Please add explanation.

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 4]
```diff
@@ -120,25 +126,32 @@ impl PointerTargetRegistry {
...
+    /// # Deprecated
+    /// This API is deprecated. Instead, use the display object's event API. For example, to get an
+    /// FRP endpoint for mouse event, you can use the [`crate::display::Object::on_event`]
+    /// function.
+    pub frp_deprecated:  enso_frp::io::Mouse_DEPRECATED,
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/scene.rs]: This structure became a bit convoluted. The responsibilities are not clear. It creates (and thus maintains) `target`, but never set the value. `hover_rgba` does not keep RGBA, as one could expect, only the target ID encoded as vector4. Again, it's created here, but never set nor read in this structure. The lack of documentation does not help.

There should be clear information, that `target` may be from a few frames ago (as it's updated from the pixel read pass).

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 5]
```diff
@@ -164,39 +180,71 @@ impl Mouse {
...
                 if display_mode.get().allow_mouse_events() {
-                    frp.up.emit(event.button())
+                    let button = event.button();
+                    frp_deprecated.up.emit(button);
+
+                    let current_target = target.get();
+                    if let Some(last_target) = last_pressed_elem.borrow_mut().remove(&button) {
+                        pointer_target_registry.with_mouse_target(last_target, |t, d| {
+                            t.emit_mouse_release(button);
+                            d.emit_event(event.clone().unchecked_convert_to::<mouse::Release>());
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/scene.rs:226]: Add comment why doing so unchecked is safe.

File: lib/rust/ensogl/core/src/display/shape.rs [hunk 6]
```diff
@@ -9,6 +9,7 @@ pub mod compound;
 pub mod constants;
 pub mod primitive;
 
+pub use compound::rectangle::Rectangle;
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/shape.rs:17]: Why is `Rectangle` re-exported, but `Circle` and `RoundedRectangle` are not?

File: lib/rust/ensogl/examples/vector-editor/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
+//! Example scene showing the usage of built-in high-level shapes.
```
Comment (farmaazon) [inline lib/rust/ensogl/examples/vector-editor/src/lib.rs]: Copied docs. The file seems to be unrelated to PR name, nor it's mentioned in the description. And I guess if we want to merge this file to develop, we must mark it as "TODO still in progress fixme".

File: lib/rust/frp/src/io/keyboard.rs [hunk 8]
```diff
@@ -440,17 +439,14 @@ pub struct DomBindings {
...
-            Listener::new_key_up(current_event.handler(
-                f!((event:&KeyboardEvent) keyboard.source.up.emit(KeyWithCode::from(event))),
-            ));
-        let blur = Listener::new_blur(
-            current_event.handler(f_!(keyboard.source.window_defocused.emit(()))),
         );
+        let key_up = Listener::new_key_up(
+            f!((event:&KeyboardEvent) keyboard.source.up.emit(KeyWithCode::from(event))),
+        );
+        let blur = Listener::new_blur(f_!(keyboard.source.window_defocused.emit(())));
```
Comment (farmaazon) [inline lib/rust/frp/src/io/keyboard.rs:449]: Why do we not set current_event anymore?

**Review by farmaazon — CHANGES_REQUESTED** [review]

**Comment by farmaazon** [comment]
Plus, the PR mentions the capturing phase, but I don't see how it could work, given that we emit mouse events directly on targets,

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 5 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/scene.rs:226]: As agreed via phone, we are OK with it as it has doc comments.

File: lib/rust/ensogl/core/src/display/shape.rs (hunk 6 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/shape.rs:17]: As agreed via phone - not a blocker (with comment why) but lets try to fix it.

File: lib/rust/frp/src/io/keyboard.rs (hunk 8 above)
Comment (wdanilo) [reply lib/rust/frp/src/io/keyboard.rs:449]: As agreed via phone, it's OK as the JsEvent was removed.

**Review by farmaazon — APPROVED** [review]

File: lib/rust/ensogl/core/src/control/io/mouse/event.rs (hunk 2 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/control/io/mouse/event.rs]: `offset_y` was broken. I'm removing it for now as no one uses it. If we need it, it needs to be implemented in a more complex way with help of display object events.

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 4 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/scene.rs]: 1. It uses `target` in the event listeners and manages the event listeners, so it kind of makes sense to store `target` here, but maybe we can find a better place.
2. `hover_rbga` - this is an old code, but this is a good catch. I renamed it to `pointer_target_encoded`.
3. Added docs to the above struct.

File: lib/rust/ensogl/core/src/display/shape.rs (hunk 6 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/shape.rs:17]: It's complex. I was trying to refactor it, but it's really complex and needs a lot of changes. This needs to be done one day. I re-exported `RoundedRectangle` and added comment about why we are not re-exporting `Circle` for now.

**Comment by wdanilo** [comment]
Regarding, `Plus, the PR mentions the capturing phase, but I don't see how it could work, given that we emit mouse events directly on targets,` - the bubbling/capturing phases are implemented for some time in display objects already and they work OK :)
